<a href="https://colab.research.google.com/github/princessivy/Analyzing_CSGO_SkinPrices/blob/main/CSGO_ML_Model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Imports

In [1]:
!pip install category_encoders --quiet

     |████████████████████████████████| 86 kB 2.3 MB/s 


In [1]:
import pandas as pd
import numpy as np
from google.colab import drive
from pathlib import Path

'''
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler, StandardScaler
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import LabelEncoder
from category_encoders import TargetEncoder
from category_encoders import CatBoostEncoder
from category_encoders import CountEncoder
import xgboost as xgb
from sklearn.metrics import mean_absolute_error'''

'\nfrom sklearn.linear_model import LinearRegression\nfrom sklearn.linear_model import LinearRegression\nfrom sklearn.model_selection import train_test_split\nfrom sklearn import metrics\nfrom sklearn.compose import ColumnTransformer\nfrom sklearn.pipeline import Pipeline\nfrom sklearn.impute import SimpleImputer\nfrom sklearn import preprocessing\nfrom sklearn.preprocessing import MinMaxScaler, StandardScaler\nfrom sklearn.preprocessing import OneHotEncoder\nfrom sklearn.preprocessing import LabelEncoder\nfrom category_encoders import TargetEncoder\nfrom category_encoders import CatBoostEncoder\nfrom category_encoders import CountEncoder\nimport xgboost as xgb\nfrom sklearn.metrics import mean_absolute_error'

# Daten einlesen

In [4]:
data = pd.read_csv(r'/content/drive/MyDrive/SA_BA/Data/allitems_cleaned_4.csv')

In [5]:
y = data['salePrice']
X = data.drop('salePrice', 1)

train_X, val_X, train_y, val_y = train_test_split(X, y, random_state = 0)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  


In [6]:
model = LinearRegression()

#### CatBoostEncoder

In [7]:
column_trans = ColumnTransformer(transformers=[
        ('scaler', StandardScaler(), ['lock', 'wear', 'exterior', 'rarity', 'stickers', 'souvenir', 'stattrak', 'colorfulness', 'cats_colorfulness']),
        ('targ', CatBoostEncoder(), ['family', 'version', 'subCategory']),
        ('ohe', OneHotEncoder(handle_unknown='ignore'), ['category', 'quality', 'collection'])], remainder = 'drop', verbose_feature_names_out=False)

pipeline = Pipeline(steps=[('t', column_trans), ('m',model)])

pipeline.fit(train_X, train_y.ravel())

score = pipeline.score(val_X, val_y.ravel())
score

0.5491667607874895

#### TargetEncoder

In [8]:
column_trans = ColumnTransformer(transformers=[
        ('scaler', StandardScaler(), ['lock', 'wear', 'exterior', 'rarity', 'stickers', 'souvenir', 'stattrak', 'colorfulness', 'cats_colorfulness']),
        ('targ', TargetEncoder(), ['family', 'version', 'subCategory']),
        ('ohe', OneHotEncoder(handle_unknown='ignore'), ['category', 'quality', 'collection'])], remainder = 'drop', verbose_feature_names_out=False)

pipeline = Pipeline(steps=[('t', column_trans), ('m',model)])

pipeline.fit(train_X, train_y.ravel())

score = pipeline.score(val_X, val_y.ravel())
score

0.5559585889347987

#### CountEncoder

In [9]:
column_trans = ColumnTransformer(transformers=[
        ('scaler', StandardScaler(), ['lock', 'wear', 'exterior', 'rarity', 'stickers', 'souvenir', 'stattrak', 'colorfulness', 'cats_colorfulness']),
        ('targ', CountEncoder(), ['family', 'version', 'subCategory']),
        ('ohe', OneHotEncoder(handle_unknown='ignore'), ['category', 'quality', 'collection'])], remainder = 'drop', verbose_feature_names_out=False)

pipeline = Pipeline(steps=[('t', column_trans), ('m',model)])

pipeline.fit(train_X, train_y.ravel())

score = pipeline.score(val_X, val_y.ravel())
score

0.23805258300534304

#### OneHotEncoder

StandardScaler vs. MinMaxScaler

In [11]:
# StandardScaler
column_trans = ColumnTransformer(transformers=[
        ('scaler', StandardScaler(), ['lock', 'wear', 'exterior', 'rarity', 'stickers', 'souvenir', 'stattrak', 'colorfulness', 'cats_colorfulness']),
        ('ohe', OneHotEncoder(handle_unknown='ignore'), ['family', 'version', 'subCategory','category', 'quality', 'collection'])], remainder = 'drop', verbose_feature_names_out=False)

pipeline = Pipeline(steps=[('t', column_trans), ('m',model)])

pipeline.fit(train_X, train_y)

score = pipeline.score(val_X, val_y.ravel())
score

0.61777241341272

In [12]:
# MinMaxScaler
column_trans = ColumnTransformer(transformers=[
        ('scaler', MinMaxScaler(), ['lock', 'wear', 'exterior', 'rarity', 'stickers', 'souvenir', 'stattrak', 'colorfulness', 'cats_colorfulness']),
        ('ohe', OneHotEncoder(handle_unknown='ignore'), ['family', 'version', 'subCategory','category', 'quality', 'collection'])], remainder = 'drop', verbose_feature_names_out=False)

pipeline = Pipeline(steps=[('t', column_trans), ('m',model)])

pipeline.fit(train_X, train_y)

score = pipeline.score(val_X, val_y.ravel())
score

0.6177718992039558

#### Ohne Pipe

In [13]:
data.columns

Index(['family', 'lock', 'version', 'salePrice', 'category', 'subCategory',
       'pattern', 'wear', 'exterior', 'quality', 'rarity', 'collection',
       'stickers', 'souvenir', 'stattrak', 'colorfulness',
       'cats_colorfulness'],
      dtype='object')

In [14]:
num = ['lock', 'wear', 'exterior', 'rarity', 'stickers', 'souvenir', 'stattrak', 'colorfulness', 'cats_colorfulness']
kat = ['family', 'version', 'category', 'subCategory', 'pattern', 'quality', 'collection']

In [15]:
y = data['salePrice']
X = data.drop('salePrice',1)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  


In [16]:
print(y.shape)
print(X.shape)

(132807,)
(132807, 16)


In [17]:
train_X, val_X, train_y, val_y = train_test_split(X, y, train_size=0.8, random_state = 0)

In [18]:
train_X_num = train_X[['lock', 'wear', 'exterior', 'rarity', 'stickers', 'souvenir', 'stattrak', 'colorfulness', 'cats_colorfulness']].copy()
val_X_num = val_X[['lock', 'wear', 'exterior', 'rarity', 'stickers', 'souvenir', 'stattrak', 'colorfulness', 'cats_colorfulness']].copy()

train_X_kat = train_X[['family', 'version', 'category', 'subCategory', 'pattern', 'quality', 'collection']].copy()
val_X_kat = val_X[['family', 'version', 'category', 'subCategory', 'pattern', 'quality', 'collection']].copy()

In [19]:
# Num-Features
scaler = StandardScaler()

train_X_num_standardized = pd.DataFrame(scaler.fit_transform(train_X_num), columns=train_X_num.columns, index=train_X_num.index)
val_X_num_standardized = pd.DataFrame(scaler.transform(val_X_num), columns=train_X_num.columns, index=val_X_num.index)

train_X_num_standardized.head()

lock      wear  exterior    rarity  stickers  souvenir  stattrak  \
85306 -0.530358  0.447583 -0.147101 -0.984476 -0.466068 -0.261207 -0.334161   
56257 -0.530358  0.661525 -0.147101  0.771054 -0.466068 -0.261207 -0.334161   
26196 -0.530358 -0.283379 -0.147101  0.185877 -0.466068 -0.261207 -0.334161   
64356  1.885518 -1.432391  1.610914  0.771054 -0.466068 -0.261207 -0.334161   
5161  -0.530358 -0.325797 -0.147101 -0.399299  2.145609  3.828381 -0.334161   

       colorfulness  cats_colorfulness  
85306     -0.890150          -0.639357  
56257      0.027157           0.733712  
26196      0.787394           0.733712  
64356     -0.302083          -0.639357  
5161       0.318196           0.733712

In [20]:
# Kat-Features
one_hot_encoder = OneHotEncoder(handle_unknown='ignore', sparse=False)
one_hot_cols_train = pd.DataFrame(one_hot_encoder.fit_transform(train_X_kat[['family', 'version', 'category', 'subCategory', 'quality', 'collection']]))
one_hot_cols_valid = pd.DataFrame(one_hot_encoder.transform(val_X_kat[['family', 'version', 'category', 'subCategory', 'quality', 'collection']]))

In [21]:
one_hot_cols_train.index = train_X_kat.index
one_hot_cols_valid.index = val_X_kat.index

In [22]:
# Num + Kat
one_hot_X_train = pd.concat([train_X_num_standardized, one_hot_cols_train], axis=1)
one_hot_X_valid = pd.concat([val_X_num_standardized, one_hot_cols_valid], axis=1)

In [23]:
one_hot_X_valid.head()

lock      wear  exterior    rarity  stickers  souvenir  stattrak  \
88181  -0.530358  0.052156 -0.147101 -0.984476  2.145609  3.828381 -0.334161   
81505  -0.530358  0.768639 -1.026109  0.771054 -0.466068 -0.261207 -0.334161   
95553  -0.530358 -0.936186  0.731906  1.356231 -0.466068 -0.261207 -0.334161   
43790   1.885518 -0.673714  0.731906  1.941408  2.145609 -0.261207 -0.334161   
121252 -0.530358  1.743441 -1.905116 -0.984476 -0.466068 -0.261207 -0.334161   

        colorfulness  cats_colorfulness    0  ...  1065  1066  1067  1068  \
88181       0.135788           0.733712  0.0  ...   0.0   0.0   0.0   0.0   
81505      -0.672963          -0.639357  0.0  ...   0.0   0.0   0.0   0.0   
95553       3.351222           3.479852  0.0  ...   0.0   0.0   0.0   0.0   
43790       1.986681           2.106782  0.0  ...   0.0   0.0   0.0   0.0   
121252     -0.792877          -0.639357  0.0  ...   0.0   0.0   0.0   0.0   

        1069  1070  1071  1072  1073  1074  
88181    0.0   0.0   0.0   0.0   0.0   0.0  
81505    0.0   0.0   0.0   0.0   0.0   0.0  
95553    0.0   0.0   0.0   0.0   0.0   0.0  
43790    0.0   0.0   0.0   0.0   0.0   0.0  
121252   0.0   0.0   0.0   0.0   0.0   0.0  

[5 rows x 1084 columns]

In [24]:
model = LinearRegression()

In [25]:
model.fit(one_hot_X_train, train_y)

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:1692: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['int', 'str']. An error will be raised in 1.2.
  FutureWarning,


LinearRegression()

In [26]:
score = model.score(one_hot_X_train, train_y)
score

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:1692: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['int', 'str']. An error will be raised in 1.2.
  FutureWarning,


0.6575920286516963

In [27]:
preds = model.predict(one_hot_X_valid)

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:1692: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['int', 'str']. An error will be raised in 1.2.
  FutureWarning,


In [29]:
mean_absolute_error(val_y, preds)

151235801.31242177

In [30]:
coefficients = pd.concat([pd.DataFrame(one_hot_X_train.columns),pd.DataFrame(np.transpose(model.coef_))], axis = 1)
coefficients

0             0
0         lock -5.361657e+00
1         wear -7.393402e-01
2     exterior  7.477862e+00
3       rarity -5.607323e+01
4     stickers  1.859519e+00
...        ...           ...
1079      1070  9.601004e+12
1080      1071  7.865986e+12
1081      1072  9.601004e+12
1082      1073  7.865986e+12
1083      1074  7.865986e+12

[1084 rows x 2 columns]

# Real

#Imports

In [77]:
import pandas as pd
import numpy as np
from google.colab import drive
from pathlib import Path

from xgboost import XGBRegressor

# Daten einlesen

In [78]:
data = pd.read_csv(r'/content/drive/MyDrive/SA_BA/Data/allitems_ready_for_model.csv')

In [79]:
data

salePrice  salePricelog  suggestedPrice  ext_Battle-Scarred  \
0         399.13      5.989287          397.21                   0   
1         362.44      5.892859          442.03                   0   
2         400.00      5.991465          397.21                   0   
3         420.73      6.041991          397.21                   0   
4         355.00      5.872118          397.21                   0   
...          ...           ...             ...                 ...   
27455      71.84      4.274441           70.72                   0   
27456      73.42      4.296196           70.72                   0   
27457      75.00      4.317488           70.72                   0   
27458      75.00      4.317488           70.72                   0   
27459      86.62      4.461531           70.72                   0   

       ext_Factory New  ext_Field-Tested  ext_Minimal Wear  ext_Well-Worn  \
0                    1                 0                 0              0   
1                    1                 0                 0              0   
2                    1                 0                 0              0   
3                    1                 0                 0              0   
4                    1                 0                 0              0   
...                ...               ...               ...            ...   
27455                0                 1                 0              0   
27456                0                 1                 0              0   
27457                0                 1                 0              0   
27458                0                 1                 0              0   
27459                0                 1                 0              0   

       rar_Classified  rar_Covert  ...  col_The Wildfire Collection  \
0                   0           1  ...                            0   
1                   0           1  ...                            0   
2                   0           1  ...                            0   
3                   0           1  ...                            0   
4                   0           1  ...                            0   
...               ...         ...  ...                          ...   
27455               0           0  ...                            0   
27456               0           0  ...                            0   
27457               0           0  ...                            0   
27458               0           0  ...                            0   
27459               0           0  ...                            0   

       col_The Winter Offensive Collection  col_The X-Ray Collection  \
0                                        0                         0   
1                                        0                         0   
2                                        0                         0   
3                                        0                         0   
4                                        0                         0   
...                                    ...                       ...   
27455                                    0                         0   
27456                                    0                         0   
27457                                    0                         0   
27458                                    0                         0   
27459                                    0                         0   

       col_The eSports 2014 Summer Collection  lock      wear  stickers  \
0                                           0     0  0.033243         0   
1                                           0     1  0.025035         0   
2                                           0     0  0.006977         0   
3                                           0     0  0.006662         0   
4                                           0     1  0.062841         0   
...                                       ...   ...       ...       ...  

In [80]:
data['suggestedPricelog'] = np.log(data['suggestedPrice'])

# Modell

## Splits

In [81]:
from sklearn.model_selection import train_test_split

In [82]:
y = data['salePrice']
X = data.drop(['salePrice', 'salePricelog'], 1)

train_X, val_X, train_y, val_y = train_test_split(X, y, random_state = 0)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  


In [83]:
# suggestedPricelog sichern
train_X_sugPr = train_X.copy()
val_X_sugPr = val_X.copy()

In [84]:
# suggestedPricelog aus normalen Daten löschen
train_X = train_X.drop(['suggestedPricelog', 'suggestedPrice'], 1)
val_X = val_X.drop(['suggestedPricelog', 'suggestedPrice'], 1)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  This is separate from the ipykernel package so we can avoid doing imports until


In [85]:
print(train_X.shape)
print(val_X.shape)

(20595, 530)
(6865, 530)


## Modellaufbau

In [32]:
import xgboost as xgb
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import MinMaxScaler, StandardScaler
from sklearn.model_selection import RandomizedSearchCV
from sklearn.pipeline import Pipeline

In [221]:
xg_reg = xgb.XGBRegressor(objective = 'reg:tweedie', booster='gblinear', eval_metric='tweedie-nloglik@1.2',tweedie_variance_power=1.1, random_state=42)

In [93]:
scaler = StandardScaler()

In [94]:
train_X_standardized = pd.DataFrame(scaler.fit_transform(train_X), columns=train_X.columns, index=train_X.index)
val_X_standardized = pd.DataFrame(scaler.transform(val_X), columns=train_X.columns, index=val_X.index)

In [222]:
xg_reg.fit(train_X_standardized, train_y)

XGBRegressor(booster='gblinear', eval_metric='tweedie-nloglik@1.2',
             objective='reg:tweedie', random_state=42,
             tweedie_variance_power=1.1)

In [223]:
preds = xg_reg.predict(val_X_standardized)

In [224]:
from sklearn.metrics import mean_absolute_error
mean_absolute_error(val_y, preds)

53.266328239625764

In [207]:
score = xg_reg.score(val_X_standardized, val_y)
score

0.7347161859156056

In [59]:
mean_absolute_error(val_y, val_X_sugPr['suggestedPrice'])

37.41654916241806

#### Hyperparametertuning mit Tree Booster

In [243]:
xg_reg_tree = xgb.XGBRegressor(colsample_bytree=0.4, eval_metric='mae', gamma=0.0, max_depth=10, n_estimators=1100, objective='reg:squarederror', random_state=42, scale_pos_weight=50)

In [226]:
scaler = StandardScaler()

In [227]:
train_X_standardized = pd.DataFrame(scaler.fit_transform(train_X), columns=train_X.columns, index=train_X.index)
val_X_standardized = pd.DataFrame(scaler.transform(val_X), columns=train_X.columns, index=val_X.index)

In [244]:
xg_reg_tree.fit(train_X_standardized, train_y)

XGBRegressor(colsample_bytree=0.4, eval_metric='mae', gamma=0.0, max_depth=10,
             n_estimators=1100, objective='reg:squarederror', random_state=42,
             scale_pos_weight=50)

In [245]:
preds = xg_reg_tree.predict(val_X_standardized)

In [246]:
from sklearn.metrics import mean_absolute_error
mean_absolute_error(val_y, preds)

50.8626326763972

In [247]:
score = xg_reg.score(val_X_standardized, val_y)
score

0.7487135764142672

In [232]:
mean_absolute_error(val_y, val_X_sugPr['suggestedPrice'])

37.41654916241806

In [238]:
hyper_grid = {
    'booster': ['gbtree', 'gblinear'],
    'n_estimators': [100, 500, 900, 1100, 1500],
    'learning_rate' : [0.05, 0.1, 0.2, 0.4, 0.6],
    'gamma' : [0.0, 0.1, 0.2, 0.4, 0.7],
    'max_depth': [2, 5, 7, 10],
    'min_child_weight': [1, 3, 5, 8],
    'colsample_bytree':[0.4, 0.5, 0.7, 0.8],
    'scale_pos_weight':[1, 30, 50, 80, 90, 100]
}

In [239]:
random_cv = RandomizedSearchCV(estimator=xg_reg_tree,
            param_distributions=hyper_grid,
            cv=5, n_iter=10,
            scoring = 'neg_mean_absolute_error',n_jobs = -1,
            verbose = 3, 
            return_train_score = True,
            random_state=42)

In [240]:
random_cv.fit(train_X_standardized, train_y)

Fitting 5 folds for each of 10 candidates, totalling 50 fits


RandomizedSearchCV(cv=5,
                   estimator=XGBRegressor(eval_metric='mae',
                                          objective='reg:squarederror',
                                          random_state=42),
                   n_jobs=-1,
                   param_distributions={'booster': ['gbtree', 'gblinear'],
                                        'colsample_bytree': [0.4, 0.5, 0.7,
                                                             0.8],
                                        'gamma': [0.0, 0.1, 0.2, 0.4, 0.7],
                                        'learning_rate': [0.05, 0.1, 0.2, 0.4,
                                                          0.6],
                                        'max_depth': [2, 5, 7, 10],
                                        'min_child_weight': [1, 3, 5, 8],
                                        'n_estimators': [100, 500, 900, 1100,
                                                         1500],
                               

In [241]:
random_cv.best_estimator_

XGBRegressor(colsample_bytree=0.4, eval_metric='mae', gamma=0.0, max_depth=10,
             n_estimators=1100, objective='reg:squarederror', random_state=42,
             scale_pos_weight=50)

In [242]:
random_cv.best_params_

{'booster': 'gbtree',
 'colsample_bytree': 0.4,
 'gamma': 0.0,
 'learning_rate': 0.1,
 'max_depth': 10,
 'min_child_weight': 1,
 'n_estimators': 1100,
 'scale_pos_weight': 50}

#### Ergebnisse mit unterschiedlichen XGB-Parametern:
* xgb.XGBRegressor(objective = 'reg:squarederror', booster='gblinear', reg_lambda=1, updater='coord_descent', random_state=42)
  * MAE: 101,72
* xgb.XGBRegressor(objective = 'reg:squarederror', booster='gbtree', random_state=42)
  * MAE: 104,70
* xgb.XGBRegressor(objective = 'reg:gamma', booster='gblinear', random_state=42)
  * MAE: 88,79
* xgb.XGBRegressor(objective = 'reg:tweedie', booster='gblinear', eval_metric='rmse', random_state=42)
  * MAE: 56,44
* xgb.XGBRegressor(objective = 'reg:tweedie', booster='gblinear', eval_metric='mae', random_state=42)
  * MAE: 56,44
* xgb.XGBRegressor(objective = 'reg:tweedie', booster='gblinear', eval_metric='tweedie-nloglik@1.2', random_state=42)
  * MAE: 56,44
* xgb.XGBRegressor(objective = 'reg:tweedie', booster='gblinear', eval_metric='tweedie-nloglik@1.2',tweedie_variance_power=1.4, random_state=42)
  * MAE:70,06
* xgb.XGBRegressor(objective = 'reg:tweedie', booster='gblinear', eval_metric='tweedie-nloglik@1.2',tweedie_variance_power=1.4, random_state=42)
  * MAE: 55,02
* xgb.XGBRegressor(objective = 'reg:tweedie', booster='gblinear', eval_metric='tweedie-nloglik@1.2',tweedie_variance_power=1.3, random_state=42)
  * MAE: 53,95
* xgb.XGBRegressor(objective = 'reg:tweedie', booster='gblinear', eval_metric='tweedie-nloglik@1.2',tweedie_variance_power=1.1, random_state=42)
  * MAE: 53,27

**Mit Hyperparametertuning:**
* xgb.XGBRegressor(colsample_bytree=0.4, eval_metric='mae', gamma=0.0, max_depth=10, n_estimators=1100, objective='reg:squarederror', random_state=42, scale_pos_weight=50)
  * MAE: 50,86